### ISSUE TITLE
Debuggable Builds [FEATURE]

### ISSUE THREAD

**Issue by briveramelo** [nrobinson2000/neopo#12]
~~Problem~~
~~I cannot use neopo to build debuggable builds for use with the vscode particle debugger~~
~~Neopo is much faster than the built-in vscode compile / flash commands from the Particle Workbench~~
~~Neopo is readily usable in CI/CD environments (thank you so much)~~
~~I use a gitlab job to automate and archive builds with neopo, but I have no way of automating builds for debuggable versions. Currently debug builds are not available from the `neopo build /path/to/project` command~~

# Problem
When I read the neopo docs, I thought it was not possible to build debug builds. The "flags" examples show how to do this, but they do not explain that the result of `neopo flags "-D DEBUG_BUILD"` will be that the build will work with the VSCode Particle Workbench Debugger. My initial reaction to this, as someone only becoming familiar with the particle ecosystem, was that it was a custom flag with only relevance to some specific project.

# Suggested Solution
`neopo build -d /path/to/project`
Use a neopo cli flag like `-d` to request a debug build

# Alternatives
* edit the docs [https://neopo.xyz/docs/full-docs](url) to indicate that `neopo flags "-D DEBUG_BUILD"` will enable compatibility with the VSCode Particle Workbench Debugger. Add this line of text beneath that example:
`Note: the 'DEBUG_ENABLED' flag will enable compatibility with the VSCode Particle Workbench Debugger`

**Comment by nrobinson2000** [comment]
Hey @briveramelo,

Thank you for using neopo. I'm glad that it has helped you with your Particle development needs.

Neopo has several options for building projects. The most common is `build`, which uses the `compile-user` target from the Workbench Makefile. However, there are also: `flash`, `flash-all` and `clean`.

To use other makefile targets you can use the `run` command, as hinted by the following in the `neopo help` output:
```
run <target> [project] [-v/q]             # Run a makefile target
```

To run a debug build you would use:

```
neopo run compile-debug /path/to/project
```

I apologize if this was [unclear in the documentation.](https://neopo.xyz/docs/full-docs#run)

The following served as a way to define `DEBUG_ENABLED` for use in application code:
```
neopo flags "-D DEBUG_ENABLED"
```

```cpp
#ifdef DEBUG_ENABLED
Serial.println("some debugging message");
#endif
```

Lately, I have used the [neopo manpage](https://github.com/nrobinson2000/neopo/blob/master/man/neopo.1) to hold the most up to date documentation [which I also host on neopo.xyz.](https://neopo.xyz/docs/neopo.html) I've been meaning to incorporate the updated documentation into the other webpage.

**Comment by briveramelo** [comment]
Yes, it has helped greatly, and it's lightning quick compared to the vscode particle workbench, so props!

What do you think about the proposals then? Adding a flag would be nice, but I do recommend at least updating the docs to clarify that the "DEBUB_ENABLED" preprocessor is not some arbitrary preprocessor, but in fact a reserved particle toolchain preprocessor needed to create debug builds.

A note like this would do the trick
`Note: the 'DEBUG_ENABLED' flag will enable compatibility with the VSCode Particle Workbench Debugger`

**Comment by nrobinson2000** [comment]
> Note: the 'DEBUG_ENABLED' flag will enable compatibility with the VSCode Particle Workbench Debugger

This is false to the best of my knowledge.

The preprocessor flag to enable debugging is `DEBUG_BUILD`, not `DEBUG_ENABLED`.
https://github.com/particle-iot/device-os/blob/develop/docs/debugging.md

In fact `DEBUG_ENABLED` isn't used anywhere in the deviceOS source. I'll admit that I had never checked for this, but you can see for yourself with:

```
$ grep -rI DEBUG_ENABLED ~/.particle/toolchains/deviceOS/2.0.0-rc.4
```

Because neopo uses the same makefile as Workbench (`~/.particle/toolchains/buildscripts/1.9.2/Makefile`), when using any of the debug targets in the makefile, `DEBUG_BUILD=y` is automatically set. An example of this would be [in the `compile-debug` target.](https://gist.github.com/nrobinson2000/0ec88acf6da554cf4806d99f92912584#file-makefile-L66)

I agree that the documentation is rough as I develop neopo in my spare time. When I revamp the documentation, I'll make sure to change that example to not contain `DEBUG` to avoid further confusion.

Have you tried using the `neopo run` command? You can use it to run any of the targets in the makefile, including the debug targets.

**Comment by briveramelo** [comment]
Oh this is bizarre. Not even the reserved `DEBUG_BUILD` flag is needed to connect via the debugger. I'm using deviceOS 2.0.0. I just deleted my target folder, ran `neopo clean && neopo flags "" && neopo build . && neopo flash .` (I am aware of the scripts feature, just did this for a quick run), and it hits breakpoints just fine. I recently got breakpoints working in CLion, so I've tested in both CLion and VSCode, and they both hit breakpoints and work as expected without any flags in the build. The strange thing is that I'm almost certain this did not work when I was first attempting, which is what prompted this entire thread.

I apologize for the distraction. The original issue must have been something else entirely.

Thank you for looking into this though
